src/server/dispatch.rs

```diff
@@ -0,0 +1,136 @@
+use super::requests::*;
+use jsonrpc_core::{self as jsonrpc};
+use server::{Request, Response, Action};
+use server::io::Output;
+use actions::InitActionContext;
+use std::sync::mpsc;
+use std::thread;
+use std::time::Duration;
+use std::fmt;
+
+lazy_static! {
+    static ref TIMEOUT: Duration = Duration::from_millis(::COMPILER_TIMEOUT);
+}
+
+/// Macro enum `DispatchRequest` packing in various similar `Request` types
+macro_rules! define_dispatch_request_enum {
+    ($($request_type:ident),*) => {
+        pub enum DispatchRequest {
+            $(
+                $request_type($request_type, Request<$request_type>, InitActionContext),
+            )*
+        }
+
+        $(
+            impl From<(Request<$request_type>, InitActionContext)> for DispatchRequest {
+                fn from((req, ctx): (Request<$request_type>, InitActionContext)) -> Self {
+                    DispatchRequest::$request_type($request_type::new(), req, ctx)
+                }
+            }
+        )*
+
+        impl DispatchRequest {
+            fn handle<O: Output>(self, out: &O) {
+                match self {
+                $(
+                    DispatchRequest::$request_type(mut var, req, ctx) => {
+                        let Request { id, params, received, .. } = req;
+                        let fallback = var.fallback_response();
+                        let timeout = var.timeout();
+
+                        let receiver = receive_from_thread(move || {
+                            if received.elapsed() >= timeout {
```

Comment:
Why do we need a timeout here?

Reply:
This timeout implements the _has it already timed out before it can begin_ check which is a big part of handling arbitrary request spam.